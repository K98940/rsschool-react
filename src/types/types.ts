export type SubmitEvents = React.MouseEvent | React.FormEvent;
export type Status = 'idle' | 'submitting' | 'success';
export type AppState = {
  query: string;
  status: Status;
};
export type ResponsePage = {
  pageNumber: number;
  pageSize: number;
  numberOfElements: number;
  totalElements: number;
  totalPages: number;
  firstPage: boolean;
  lastPage: boolean;
};
export type ResponseSortClause = {
  name: string;
  direction: 'ASC' | 'DESC';
  clauseOrder: number;
};
export type ResponseSort = {
  clauses: ResponseSortClause[];
};
export type Header = {
  uid: string;
  title: string;
};
export type EpisodeBase = {
  uid: string;
  title: string;
  titleGerman: string;
  titleItalian: string;
  titleJapanese: string;
  series: Header;
  season: Header;
  seasonNumber: number;
  episodeNumber: number;
  productionSerialNumber: number;
  featureLength: number;
  stardateFrom: number;
  stardateTo: number;
  yearFrom: number;
  yearTo: number;
  usAirDate: string;
  finalScriptDate: string;
};
export type EpisodeBaseResponse = {
  page: ResponsePage;
  sort: ResponseSort;
  episodes: EpisodeBase[];
};
export type EpisodeFull = {
  uid: string;
  title: string;
  series: {
    title: string;
  };
  season: {
    title: string;
  };
  usAirDate: string;
};
export type EpisodeFullResponse = {
  episode: EpisodeFull;
};
export type Pagination = {
  pageNumber: number;
  pageSize: number;
  numberOfElements: number;
  totalElements: number;
  totalPages: number;
  firstPage: boolean;
  lastPage: boolean;
};
export type PaginationState = {
  pagination: Pagination;
};
export type EpisodesState = {
  episodes: EpisodeBase[];
};
export type PaginationPayload = {
  payload: Pagination;
  type: string;
};
export type EpisodesPayload = {
  payload: EpisodeBase[];
  type: string;
};
