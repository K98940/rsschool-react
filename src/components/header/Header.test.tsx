import Header from './header';
import { Status } from '@/types/types';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

const onChange = vi.fn(() => 'onChange');
const onReset = vi.fn(() => 'onReset');
const onSearch = vi.fn(() => 'onSearch');
const TEST_QUERY = 'test';

describe('Header component', () => {
  const renderComponent = (query: string, status: Status = 'idle') => {
    return {
      snapshot: render(
        <Header
          query={query}
          handleQueryChange={onChange}
          handleQueryReset={onReset}
          handleSearch={onSearch}
          status={status}
        />,
      ),
      searchInput: screen.getByTestId('search-input'),
      resetButton: screen.queryByTestId('reset-btn'),
      searchButton: screen.getByTestId('search-btn'),
    };
  };

  it('Header snapshot with provided search query', () => {
    const { snapshot } = renderComponent(TEST_QUERY);
    expect(snapshot).toMatchSnapshot();
  });

  it('Header snapshot with empty search query', () => {
    const { snapshot } = renderComponent('');
    expect(snapshot).toMatchSnapshot();
  });

  it('Header snapshot with status === "submitting"', () => {
    const { snapshot } = renderComponent('');
    expect(snapshot).toMatchSnapshot();
  });

  it('click Search button should call onSearch callback in the first time', async () => {
    const { searchButton } = renderComponent(TEST_QUERY);
    const user = userEvent.setup();

    await user.click(searchButton);
    expect(onSearch).toBeCalledTimes(1);
  });

  it('press keyboard Enter button should call onSearch callbackin the second time', async () => {
    renderComponent(TEST_QUERY);
    const user = userEvent.setup();

    await user.keyboard('[Enter]');
    expect(onSearch).toBeCalledTimes(2);
  });

  it('click Reset button should call onReset callback', async () => {
    const { resetButton } = renderComponent(TEST_QUERY);
    const user = userEvent.setup();

    expect(resetButton).toBeInTheDocument();
    if (!resetButton) throw new Error();
    await user.click(resetButton);
    expect(onReset).toHaveBeenCalled();
  });

  it('typing in the search input should call onChange callback', async () => {
    const { searchInput } = renderComponent(TEST_QUERY);
    const user = userEvent.setup();

    await user.type(searchInput, 'aaa');
    expect(onChange).toHaveBeenCalledTimes(3);
  });
});
