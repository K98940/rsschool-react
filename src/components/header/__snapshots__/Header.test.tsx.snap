// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header component > Header snapshot with empty search query 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="_header_503ab9"
      >
        <div
          class="_searchContainer_503ab9"
        >
          <form
            class="_inputContainer_503ab9"
          >
            <input
              class="_input_503ab9"
              data-testid="search-input"
              placeholder="Search for the Star Trek episode"
              value=""
            />
          </form>
          <button
            class="_btnSearch_503ab9"
            data-testid="search-btn"
          >
            Search
          </button>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="_header_503ab9"
    >
      <div
        class="_searchContainer_503ab9"
      >
        <form
          class="_inputContainer_503ab9"
        >
          <input
            class="_input_503ab9"
            data-testid="search-input"
            placeholder="Search for the Star Trek episode"
            value=""
          />
        </form>
        <button
          class="_btnSearch_503ab9"
          data-testid="search-btn"
        >
          Search
        </button>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Header component > Header snapshot with provided search query 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="_header_503ab9"
      >
        <div
          class="_searchContainer_503ab9"
        >
          <form
            class="_inputContainer_503ab9"
          >
            <input
              class="_input_503ab9"
              data-testid="search-input"
              placeholder="Search for the Star Trek episode"
              value="test"
            />
            <button
              class="_btnReset_503ab9"
              data-testid="reset-btn"
              type="reset"
            >
              x
            </button>
          </form>
          <button
            class="_btnSearch_503ab9"
            data-testid="search-btn"
          >
            Search
          </button>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="_header_503ab9"
    >
      <div
        class="_searchContainer_503ab9"
      >
        <form
          class="_inputContainer_503ab9"
        >
          <input
            class="_input_503ab9"
            data-testid="search-input"
            placeholder="Search for the Star Trek episode"
            value="test"
          />
          <button
            class="_btnReset_503ab9"
            data-testid="reset-btn"
            type="reset"
          >
            x
          </button>
        </form>
        <button
          class="_btnSearch_503ab9"
          data-testid="search-btn"
        >
          Search
        </button>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Header component > Header snapshot with status === "submitting" 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="_header_503ab9"
      >
        <div
          class="_searchContainer_503ab9"
        >
          <form
            class="_inputContainer_503ab9"
          >
            <input
              class="_input_503ab9"
              data-testid="search-input"
              placeholder="Search for the Star Trek episode"
              value=""
            />
          </form>
          <button
            class="_btnSearch_503ab9"
            data-testid="search-btn"
          >
            Search
          </button>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="_header_503ab9"
    >
      <div
        class="_searchContainer_503ab9"
      >
        <form
          class="_inputContainer_503ab9"
        >
          <input
            class="_input_503ab9"
            data-testid="search-input"
            placeholder="Search for the Star Trek episode"
            value=""
          />
        </form>
        <button
          class="_btnSearch_503ab9"
          data-testid="search-btn"
        >
          Search
        </button>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
